beacon-chain/types/attestation.go

```diff
@@ -0,0 +1,99 @@
+// Package types defines the essential types used throughout the beacon-chain.
+package types
+
+import (
+	"fmt"
+
+	"github.com/golang/protobuf/proto"
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"golang.org/x/crypto/blake2b"
+)
+
+// Attestation is the primary source of load on the beacon chain, it's used to
+// attest to some parent block in the chain and block hash in a shard.
+type Attestation struct {
+	data *pb.AttestationRecord
+}
+
+// NewAttestation explicitly sets the data field of a attestation.
+// Return attestation with default fields if data is nil.
+func NewAttestation(data *pb.AttestationRecord) *Attestation {
+	if data == nil {
+		return &Attestation{
+			data: &pb.AttestationRecord{
+				Slot:                0,
+				ShardId:             0,
+				JustifiedSlot:       0,
+				JustifiedBlockHash:  []byte{},
+				ShardBlockHash:      []byte{},
+				AttesterBitfield:    []byte{},
+				ObliqueParentHashes: [][]byte{{}},
+				AggregateSig:        []uint64{},
+			},
+		}
+	}
+	return &Attestation{data: data}
+}
+
+// Proto returns the underlying protobuf data.
+func (a *Attestation) Proto() *pb.AttestationRecord {
+	return a.data
+}
+
+// Marshal encodes block object into the wire format.
+func (a *Attestation) Marshal() ([]byte, error) {
+	return proto.Marshal(a.data)
+}
+
+// Hash generates the blake2b hash of the attestation.
+func (a *Attestation) Hash() ([32]byte, error) {
```

Comment:
This isn't necessarily a suggestion for this PR, but one idea for a refactor is to tweak the constructor for `types.Block`, `types.ActiveState`, `types.CrystallizedState`, and now `types.Attestation` so that the constructor accepts the encoding, not the proto object. Since all of these objects are immutable, we can keep the encoding and return it for these `Marhsal` and `Hash` functions instead of re-encoding the proto object every time. 

The upside is simpler (calling `Hash` wouldn't require error handling) and faster code, the downside is increased memory usage. What do you think?

Reply:
Nice. I'm on board with that. That makes a lot of sense. Do you plan to do that as part of #473? if not we can open another PR for it. 

Noob question - why does using the encoding cause more mem usage? 

To answer your earlier question, I will replace `AttestationRecord` with `AttestationRecordHash` in a later PR

Reply:
I would do my suggestion in a separate PR

Each object would have both the binary encoding and the protobuf message. Right now we discard the binary encoding and only use the protobuf message. I guess this is why something like flatbuffers would be nice, since it enables access to fields directly from the encoding.

Gotcha! Separate PR SGTM.